### DISCUSSION TITLE
Does test-utils work with webgl2?

### DISCUSSION THREAD

**Discussion by jeffpeck10x** [visgl/deck.gl#6759] [Q&A]
I have some glsl files in a custom layer that are using webgl2:
```
#version 300 es
```

Prior to switching to that version, the unit tests were working. After doing so however, I get this:
```
ERROR: unsupported shader version
```

Looking through deck.gl source for answers, I found that I can do this inside `getShaders`:
```
    const transpileToGLSL100 = !isWebGL2(this.context.gl);

    return super.getShaders({
      fs,
      modules: [project32, picking],
      transpileToGLSL100,
      vs,
    });
```

Except... doing so creates more glsl compilation errors:
```
      ^^^ ERROR: 'attribute' : cannot be bool or int
```

```
      ^^^ ERROR: 'flat' : Illegal use of reserved word
```

I guess the transpiler doesn't know how to handle ints for attribute or varying variables. I would make a bug report for this, but I'm guessing this is an obscure use-case. Instead, I would like to just run the tests in a webgl2 context.

How can I force the test runner to use webgl2?

It looks like it is possible as noted here (`webgl2TestDevice`):
https://github.com/visgl/luma.gl/blob/master/modules/test-utils/src/create-test-device.ts

Is there some way to tell the test runner to use that or something similar?

**Comment by jeffpeck10x** [comment]
I found that I can do this in `jest.setup.js`:
```
globalThis.glContext = createTestContext({
  debug: true,
  height: 1,
  webgl1: false,
  webgl2: true,
  width: 1,
});
```

However, the result is `null`. So, this does not work either.

**Comment by ibgreen** [comment]
Note that tests are run on Node.js but Node.js doesn't support WebGL. We get around this in two ways:

- we install "headless-gl" but that only supports WebGL1. 
- We also run tests in a headless Chrome instance via puppeteer. This will support WebGL2.

You would need to do the same in your environment.

**Reply by jeffpeck10x** [reply]
Are there any steps to tell the deckgl test-utils to use the latter headless Chrome instance? I am unfamiliar with the best way to accomplish this.

**Comment by ibgreen** [comment]
Haven't looked at that in quite a while. there are test runner classes in `@probe.gl/test-utils`. 

https://uber-web.github.io/probe.gl/docs/api-reference/test-utils/browser-test-driver.

**Reply by jeffpeck10x** [reply]
I guess.... how would I even tell deckgl `test-utils` to use a different gl context? Is setting `globalThis.glContext` the right way to go? It  seems a little hacky, although it could work, if only I could get a valid context.
